File: src/String/299-BullsandCows/main.cpp
#include <iostream>
#include <vector>
#include <cstring>

using namespace std;

class Solution299
{
public:
  string getHint(string secret, string guess)
  {
    int n = secret.size();
    int bulls = 0, cows = 0;
    int seCnt[10], guCnt[10];
    memset(seCnt, 0, sizeof(seCnt));
    memset(guCnt, 0, sizeof(guCnt));

    for (int i = 0; i < n; i++)
    {
      // 要是对应的位置相等的话 bulls 的值就+1
      if (secret[i] == guess[i])
      {
        bulls++;
      }
      else // 要是对应位置不相等就把该位置的值的个数存下来
      {
        seCnt[secret[i] - '0']++;
        guCnt[guess[i] - '0']++;
      }
    }

    // 然后 cows 的值就是两者之间的最小值
    for (int i = 0; i <= 9; i++)
      cows += min(seCnt[i], guCnt[i]);

    return to_string(bulls) + "A" + to_string(cows) + "B";
  }
};

int main()
{
  string secret1 = "1807", guess1 = "7810";
  cout << Solution299().getHint(secret1, guess1);

  return 0;
}

File: src/String/299-BullsandCows/main.py
class Solution:
  def getHint(self, secret: str, guess: str) -> str:
    bulls = 0
    cows = 0
    cnt = [0] * 10
    n = len(secret)
    visited = [0] * n

    for i in range(n):
      if secret[i] == guess[i]:
        bulls += 1
        visited[i] = 1
      else:
        cnt[int(secret[i])] += 1

    for i in range(n):
      if visited[i] == 0 and cnt[int(guess[i])] > 0:
        cows += 1
        cnt[int(guess[i])] -= 1

    return "{}A{}B".format(str(bulls), str(cows))
